### PR TITLE
Spline fitting routines that mimic IRAF fitting

### PR DESCRIPTION
**Pull request by stscirij**
For use in defringe tasks

### CODE REVIEW

File: fit1d.py [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+    -------
+
+    fit : scipy.interpolate.fitpack2.LSQUnivariateSpline object
+        The result of the fit.  It can be used to calculate the fitted
+        values for the input x values:
+
+        fitted_values = fit(x)
+
+    """
+
```
Comment (sean-lockwood) [inline fit1d.py]:
```
if grow:  
    raise NotImplementedError()
```

File: fit1d.py [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+    order : int
+        Order of fitting fuction.  The number of knots in the spline fit is (order-1)
+
+    low_reject : float
+        Points with y values less than fit - low_reject*(rms deviation of data from fit) are rejected
+        when the fit is iterated.  Ignored if niterate = 0.
+
+    high_reject : float
+        Points with y values greater than fit + low_reject*(rms deviation of data from fit) are rejected
```
Comment (sean-lockwood) [inline fit1d.py]: `fit + high_reject`

File: fit1d.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+        fitted_values = fit(x)
+
+    """
+
+    fitted = fit_once(x, y, weights, function, order)
+    if niterate <= 0:
+        return fitted
+    npts = len(x)
+    rej_weights = np.ones(npts)
+    for iter in range(niterate):
```
Comment (sean-lockwood) [inline fit1d.py]: Best practice is probably to avoid overriding `iter`:
`for _ in range(niterate):`

File: fit1d.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+    """
+
+    if nbin == 1:
+        return x, y, weights
+    nx = len(x)
+    ny = len(y)
+    if nx != ny:
+        print("Lengths of x and y vectors different, aborting")
+        return None, None, None
+    if type(weights) == type(None):
```
Comment (sean-lockwood) [inline fit1d.py]: `if weights is None:`

File: fit1d.py [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+        weights = np.ones(nx)
+    nw = len(weights)
+    if nx != nw:
+        print("Lengths of data and weight vectors different, aborting")
+        return None, None, None
```
Comment (sean-lockwood) [inline fit1d.py]: An exception might be more explicit here:
`raise ValueError('Lengths of data and weight vectors different')`

File: fit1d.py [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+
+    naverage : int
+        The number of adjacent elements of x, y, and weight that are averaged before
+        fitting is done.  If naverage=1, no averaging is done.
+
+    function : str
+        Fitting function.  Currently only "spline1" and "spline3" are supported
+
+    order : int
+        Order of fitting fuction.  The number of knots in the spline fit is (order-1)
```
Comment (philhodge) [inline fit1d.py]: The only thing I would add to previous comments is that this description of `order` should be more like the description in `fit_once`.

File: fit1d.py [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+        This can be used to calculate the fitted values for the input x values:
+
+        fitted_values = fit(x)
+
+    """
+
+    nx = len(x)
+    if function[:6] == "spline":
+        #
+        # Number of knots is 1 fewer than the order
```
Comment (philhodge) [inline fit1d.py]: I would have expected the number of knots to be order + 1.  The reason it's order - 1 is that the scipy fitting routine requires that the knots be internal, i.e. there is no knot at either endpoint.

**Comment by sean-lockwood** [comment]
Don't forget to move it to the `stistools` subdirectory.

**Pull request merged**